cpp/src/hash/sha_hash.cuh

```diff
@@ -0,0 +1,537 @@
+{
+  uint32_t words[64];
+
+  // The 512-bit message buffer fills the first 16 words.
+  memcpy(&words[0], state.buffer, sizeof(words[0]) * 16);
+  for (int i = 0; i < 16; i++) {
+    // Convert word representation from little-endian to big-endian.
+    words[i] = swap_endian(words[i]);
+  }
+
+  // The rest of the 64 words are generated from the first 16 words.
+  for (int i = 16; i < 64; i++) {
+    uint32_t const s0 = rotate_bits_right(words[i - 15], 7) ^ rotate_bits_right(words[i - 15], 18) ^
+                        (words[i - 15] >> 3);
+    uint32_t const s1 = rotate_bits_right(words[i - 2], 17) ^ rotate_bits_right(words[i - 2], 19) ^
+                        (words[i - 2] >> 10);
+    words[i] = words[i - 16] + s0 + words[i - 7] + s1;
+  }
+
+  uint32_t A = state.hash_value[0];
+  uint32_t B = state.hash_value[1];
+  uint32_t C = state.hash_value[2];
+  uint32_t D = state.hash_value[3];
+  uint32_t E = state.hash_value[4];
+  uint32_t F = state.hash_value[5];
+  uint32_t G = state.hash_value[6];
+  uint32_t H = state.hash_value[7];
+
+  for (int i = 0; i < 64; i++) {
+    uint32_t const s1 =
+      rotate_bits_right(E, 6) ^ rotate_bits_right(E, 11) ^ rotate_bits_right(E, 25);
+    uint32_t const ch    = (E & F) ^ ((~E) & G);
+    uint32_t const temp1 = H + s1 + ch + sha256_hash_constants[i] + words[i];
+    uint32_t const s0 =
+      rotate_bits_right(A, 2) ^ rotate_bits_right(A, 13) ^ rotate_bits_right(A, 22);
+    uint32_t const maj   = (A & B) ^ (A & C) ^ (B & C);
+    uint32_t const temp2 = s0 + maj;
+
+    H = G;
+    G = F;
+    F = E;
+    E = D + temp1;
+    D = C;
+    C = B;
+    B = A;
+    A = temp1 + temp2;
+  }
+
+  state.hash_value[0] += A;
+  state.hash_value[1] += B;
+  state.hash_value[2] += C;
+  state.hash_value[3] += D;
+  state.hash_value[4] += E;
+  state.hash_value[5] += F;
+  state.hash_value[6] += G;
+  state.hash_value[7] += H;
+
+  state.buffer_length = 0;
+}
+
+/**
+ * @brief Core SHA-512 algorithm implementation. Processes a single 1024-bit chunk,
+ * updating the hash value so far. Does not zero out the buffer contents.
+ */
+template <typename hash_state>
+void __device__ inline sha512_hash_step(hash_state& state)
+{
+  uint64_t words[80];
+
+  // The 1024-bit message buffer fills the first 16 words.
+  memcpy(&words[0], state.buffer, sizeof(words[0]) * 16);
+  for (int i = 0; i < 16; i++) {
+    // Convert word representation from little-endian to big-endian.
+    words[i] = swap_endian(words[i]);
+  }
+
+  // The rest of the 80 words are generated from the first 16 words.
+  for (int i = 16; i < 80; i++) {
+    uint64_t const s0 = rotate_bits_right(words[i - 15], 1) ^ rotate_bits_right(words[i - 15], 8) ^
+                        (words[i - 15] >> 7);
+    uint64_t const s1 = rotate_bits_right(words[i - 2], 19) ^ rotate_bits_right(words[i - 2], 61) ^
+                        (words[i - 2] >> 6);
+    words[i] = words[i - 16] + s0 + words[i - 7] + s1;
+  }
```

Comment:
I'm not an expert on SHA, but there are a fair number of these magic numbers in this file. Is there any value in explicitly giving them variable names?  For example the 64 and 80 seem very explicitly tied to the 256/512 widths.

Reply:
I don't know what I would name them. I studied a few SHA implementations while writing this, and didn't find any meaningful names for these values besides things like `magic1`, `magic2`, ... Even still, I don't know that the _relationships_ between these magic values is meaningful, either. Like is every instance of the number `15` meaning the same thing or different things that both happen to be 15? I think I would keep the status quo here unless you can point to an implementation that does better.